### PR TITLE
Changes to Menu Item Spacing

### PR DESCRIPTION
**Pull request by stonerl**
- Instead of killing apps we terminate() them gracefully. Only if they do not terminate within 3 second, we forceTerminate() them.
- Disable buttons while restarting apps
- Draw progress circle
- Make Ice settings window always stay on top

### CODE REVIEW

File: Ice/Settings/SettingsWindow.swift [hunk 1]
```diff
@@ -16,9 +16,28 @@ struct SettingsWindow: Scene {
...
+        DispatchQueue.main.async {
+            if let window = view.window {
+                self.onReceiveWindow(window)
+            }
+        }
+        return view
+    }
+
+    func updateNSView(_ nsView: NSView, context: Context) {}
+}
```
Comment (jordanbaird) [inline Ice/Settings/SettingsWindow.swift:43]:
```suggestion
```

File: Ice/Settings/SettingsWindow.swift [hunk 2]
```diff
@@ -16,9 +16,28 @@ struct SettingsWindow: Scene {
                 .onAppear(perform: onAppear)
                 .environmentObject(appState)
                 .environmentObject(appState.navigationState)
+                .background(WindowAccessor { window in
+                    window.level = .floating
+                })
```
Comment (jordanbaird) [inline Ice/Settings/SettingsWindow.swift:21]: We have a custom modifier, `View.readWindow(onChange:)`, so let's use that here instead.

```suggestion
                .readWindow { window in
                    window.level = .floating
                }
```

File: Ice/Settings/SettingsPanes/GeneralSettingsPane.swift [hunk 3]
```diff
@@ -12,54 +12,55 @@ struct GeneralSettingsPane: View {
...
-            }
-            Section {
-                spacingOptions
-            }
+            Section { launchAtLogin }
+            Section { iceIconOptions }
+            Section { useIceBar }
+            Section { showOnClick; showOnHover; showOnScroll }
+            Section { autoRehideOptions }
+            Section { spacingOptions }
```
Comment (jordanbaird) [inline Ice/Settings/SettingsPanes/GeneralSettingsPane.swift:63]: Let's keep the original formatting here.

```suggestion
            Section {
                launchAtLogin
            }
            Section {
                iceIconOptions
            }
            Section {
                useIceBar
            }
            Section {
                showOnClick
                showOnHover
                showOnScroll
            }
            Section {
                autoRehideOptions
            }
            Section {
                spacingOptions
            }
```

**Review by jordanbaird — CHANGES_REQUESTED** [review]
Great PR! This should make the spacing feature a lot more robust. If you can make the suggested changes, this should be ready to merge.

**Comment by stonerl** [comment]
Should be good now.

**Review by jordanbaird — CHANGES_REQUESTED** [review]
Getting some build errors and warnings involving concurrency. Can you resolve those?

**Comment by stonerl** [comment]
Did you pull the latest commit? I had to force push.

**Comment by jordanbaird** [comment]
Yes. These are all in MenuBarItemSpacingManager.swift. I'm using Xcode 15.4 on macOS 14.6.1.

CleanShot 2024-08-09 at 03 19 27@2x

**Comment by stonerl** [comment]
Hmm. I'm using Xcode 16 beta 5 on macOS 15 beta 5.

That could be the reason why I don't get these errors…

**Comment by jordanbaird** [comment]
That would be my assumption. Could be either a compiler regression, or it could be that these errors are false positives and the latest compiler fixed them. Either way, we have to make sure it builds with the latest stable Xcode release. I can fix the errors if you want, or I'm good to wait if you would prefer to.

**Comment by stonerl** [comment]
I think it is faster if you fix them. Xcode 15 does not run on macOS 15 and I would have to set up a macOS 14.6 VM in order to do so myself.

**Comment by jordanbaird** [comment]
Want to make sure it builds and works correctly on your machine with the changes I made?

**Review by jordanbaird — APPROVED** [review]

**Comment by stonerl** [comment]
works

**Pull request merged**